### PR TITLE
Add FastMTP multi-token prediction speculator

### PR DESCRIPTION
**Pull request by rahul-tuli**
## Motivation

Multi-token prediction (MTP) heads are now standard in frontier models (DeepSeek-V3,
Qwen3, TencentBAC MiMo), but each ships in its own bespoke checkpoint format with
no standardized inference or fine-tuning interface. Loading them today requires either
the originating codebase or a custom conversion script.

This PR adds `FastMTPSpeculator` — a Hugging Face-compatible MTP module in the
speculators format. It provides:

1. A generic MTP layer definition suitable for fine-tuning from scratch or continued
   training on top of converted upstream checkpoints.
2. A registered config (`"mtp"`) and model (`FastMTPSpeculator`) that integrate with
   `SpeculatorModel.from_pretrained` and the existing vLLM deployment path.
3. Weight key normalization utilities to support the conversion pipeline that maps
   upstream checkpoint formats (MiMo, Qwen3-Next, DeepSeek) into the speculators format.

> **Note:** Upstream checkpoints require a conversion step before they can be loaded
> directly by speculators. The conversion utilities are part of a separate script;
> this PR defines the target format and module.

### `FastMTPConfig` — configuration

Pydantic config registered as `"mtp"` under `SpeculatorModelConfig`:

- `num_speculative_steps` (default: 3, range: 1–10) — tokens predicted per forward pass
- `num_nextn_predict_layers` (always 1) — validated at construction
- `mtp_loss_step_weights` (default: `[0.51, 0.31, 0.18]`) — per-step loss weighting for training
- Standard transformer dims (`hidden_size`, `intermediate_size`, `num_attention_heads`, etc.)
- `transformer_config: PretrainedConfig` — architecture config for MTP layer components

### `FastMTPLayer` — single MTP decoder layer

A single transformer layer that fuses verifier hidden states with token embeddings to
predict the next token:

```mermaid
flowchart TD
    H["hidden_states\n(from verifier)"] --> HN["hidden_layernorm"]
    E["embed_tokens(input_ids)"] --> TN["token_layernorm"]
    HN --> CAT["Concatenate\n[hidden ‖ embed]"]
    TN --> CAT
    CAT --> PROJ["input_proj\n(2H → H, no bias)"]
    PROJ --> LN1["input_layernorm"]
    LN1 --> ATT["Self-Attention\n+ RotaryEmb"]
    ATT --> RES1["+ residual"]
    RES1 --> LN2["post_attention_layernorm"]
    LN2 --> MLP["MLP / SparseMoeBlock"]
    MLP --> RES2["+ residual"]
    RES2 --> LN3["final_layernorm"]
    LN3 --> OUT["layer output"]
```

### `FastMTPSpeculator` — multi-step predictor

Performs `num_speculative_steps` auto-regressive passes, each using the previous
step's argmax output as the next input:

```
for step in range(num_speculative_steps):
    layer_out = FastMTPLayer(input_ids, hidden_states)
    logits    = lm_head(layer_out)          # shared with verifier
    loss     += weight[step] * CE(logits, labels[step+1:])
    input_ids = argmax(logits)              # feeds next step
    hidden_states = layer_out
```

`embed_tokens` and `lm_head` are shared with the verifier model and frozen during
training.

### Architecture-agnostic component registry

`FastMTPComponents` maps architecture names to their concrete module classes:

| Architecture | Attention | MLP | Norm | MoE |
|---|---|---|---|---|
| `mimo` | `Qwen2Attention` | `Qwen2MLP` | `Qwen2RMSNorm` | No |
| `qwen3_next` | `Qwen3NextAttention` | `Qwen3NextSparseMoeBlock` | `Qwen3NextRMSNorm` | Yes |

New architectures can be added by registering a new entry — no changes to the forward pass.

### Checkpoint key normalization

`normalize_weight_key` maps weight keys from upstream checkpoint formats into the
speculators key namespace, which is used by the conversion pipeline:

| Source format | Example key | Speculators key |
|---|---|---|
| `qwen3_next` | `mtp.layers.0.self_attn.q_proj.weight` | `model.mtp_layers.0.self_attn.q_proj.weight` |
| `qwen3_next` | `mtp.fc.weight` | `model.mtp_layers.0.input_proj.weight` |
| `qwen3_next` | `mtp.norm.weight` | `model.mtp_layers.0.final_layernorm.weight` |
| `deepseek` | `mtp.<component>` | `model.mtp_layers.0.<component>` |

### `MiMoConfig` — AutoConfig shim

Registers `model_type="mimo"` with `transformers.CONFIG_MAPPING` so TencentBAC
MiMo checkpoints can be loaded via `AutoConfig` without requiring their patched
transformers fork.

## Test Checkpoints

Integration tests run against two HuggingFace-hosted test checkpoints that have
already been converted to the speculators format:

- [`inference-optimization/test_tencentbac_fastmtp`](https://huggingface.co/inference-optimization/test_tencentbac_fastmtp) — MiMo/TencentBAC (Qwen2-based)
- [`inference-optimization/test_qwen3_next_mtp`](https://huggingface.co/inference-optimization/test_qwen3_next_mtp) — Qwen3-Next (MoE)

## Testing

```bash
# Integration tests (requires GPU)
CUDA_VISIBLE_DEVICES=3 pytest tests/integration/models/test_fast_mtp_model.py -v --tb=short
```

Tests validate for both checkpoint formats:
- `from_pretrained` loads `FastMTPSpeculator` with the correct config type
- Forward pass completes with mock `embed_tokens` / `lm_head` in detached mode
- `logits_list` has length == `num_speculative_steps`
- Each logits tensor has shape `(batch, seq_len, vocab_size)`
- No NaN or Inf values in any output tensor

## Files Changed

| File | Change |
|---|---|
| `src/speculators/models/fast_mtp/config.py` | `FastMTPConfig` with Pydantic validators |
| `src/speculators/models/fast_mtp/core.py` | `FastMTPLayer` + `FastMTPSpeculator` |
| `src/speculators/models/fast_mtp/model_definitions.py` | `FastMTPComponents` registry (mimo, qwen3_next) |
| `src/speculators/models/fast_mtp/mimo_config.py` | `MiMoConfig` AutoConfig shim |
| `src/speculators/models/fast_mtp/__init__.py` | Package exports |
| `src/speculators/models/__init__.py` | Auto-import for registry |
| `src/speculators/model.py` | Handle `speculators_config=None` edge case |
| `tests/integration/models/test_fast_mtp_model.py` | Parametrized integration tests |

### CODE REVIEW

File: src/speculators/models/fast_mtp/core.py [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+            config=config,
+            verifier=verifier,
+            verifier_attachment_mode=verifier_attachment_mode or "detached",
+        )
+
+        self.hidden_size = config.hidden_size
+        self.vocab_size = config.vocab_size
+        self.num_speculative_steps = config.num_speculative_steps
+
+        self.mtp_layers = nn.ModuleList([FastMTPLayer(config, layer_idx=0)])
```
Comment (brian-dellabetta) [inline src/speculators/models/fast_mtp/core.py]: why make this a module list? it looks like you only ever access index 0. is it always length 1?

File: src/speculators/models/fast_mtp/core.py [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+        device = input_ids.device
+
+        all_logits = []
+        total_loss = torch.tensor(0.0, device=device) if labels is not None else None
+        metrics = {}
+
+        current_hidden = hidden_states
+        current_input_ids = input_ids
+
+        for step in range(self.num_speculative_steps):
```
Comment (brian-dellabetta) [inline src/speculators/models/fast_mtp/core.py]: num_speculative_steps here is the same as number of tokens to predict, right?

**Review by brian-dellabetta — COMMENTED** [review]
interesting! following this for the most part. just a couple questions

File: src/speculators/models/fast_mtp/config.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    :param rms_norm_eps: Epsilon for RMS normalization
+    """
+
+    speculators_model_type: Literal["mtp"] = "mtp"
+    architectures: list[str] = Field(
+        default_factory=lambda: ["FastMTPSpeculator"],
+        description="Model architectures that can load these weights",
+    )
+
+    transformer_config: PretrainedConfig = Field(
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/config.py]: Maybe use `transformer_layer_config` for consistency with Eagle3

File: src/speculators/models/fast_mtp/config.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        default_factory=Qwen2Config,
+        description=(
+            "Configuration for the transformer architecture (e.g., Qwen2Config)"
+        ),
+    )
+
+    num_speculative_steps: int = Field(
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/config.py]: I think we're currently using the proposal_methods `GreedyTokenProposalConfig` to store this.

File: src/speculators/models/fast_mtp/config.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        default=3,
+        ge=1,
+        le=10,
+        description="Number of future tokens to predict per forward pass",
+    )
+
+    num_nextn_predict_layers: int = Field(
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/config.py:66]: Maybe use `num_hidden_layers` on the `transformer_layer_config`?

File: src/speculators/models/fast_mtp/config.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        default=1,
+        description="Number of MTP layers (currently only 1 is supported)",
+    )
+
+    mtp_loss_step_weights: list[float] = Field(
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/config.py]: This is a training time setting and doesn't effect model usage. I don't think we should include this in the model config.

File: src/speculators/models/fast_mtp/config.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        default=[0.51, 0.31, 0.18],
+        description="Loss weights for each prediction step",
+    )
+
+    hidden_size: int = Field(
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/config.py]: These fields are all duplicated on the `transformer_layer_config`. We should just have a single source of truth for them.

File: src/speculators/models/fast_mtp/config.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        default=32,
+        description="Number of attention heads",
+    )
+
+    num_key_value_heads: int = Field(
+        default=8,
+        description="Number of key-value heads for grouped query attention",
+    )
+
+    vocab_size: int = Field(
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/config.py]: Maybe `draft_vocab_size` if we want fastmtp to support vocab mappings

File: src/speculators/models/fast_mtp/core.py [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+        self.lm_head: nn.Linear | None = None
+
+        if verifier is not None and verifier_attachment_mode != "detached":
+            self.attach_verifier(verifier, mode=verifier_attachment_mode)
+
+    def attach_verifier(
+        self,
+        verifier: str | os.PathLike | PreTrainedModel,
+        mode: Literal["full", "train_only"] | None = None,
+    ):
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/core.py]: The `attach_verifier` system is deprecated. It doesn't scale to training workflow as it requires loading the entire verifier model, which can be significantly larger than the draft model.

File: src/speculators/models/fast_mtp/core.py [hunk 10]
```diff
@@ -0,0 +1,514 @@
...
+        hidden_states: torch.Tensor,
+        attention_mask: torch.Tensor | None = None,
+        position_ids: torch.Tensor | None = None,
+        labels: torch.Tensor | None = None,
+        return_dict: bool = True,
+        **kwargs,
+    ) -> dict[str, Any] | tuple:
+        """Forward pass for FastMTP multi-token prediction.
+
+        :param input_ids: Token IDs [batch, seq_len]
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/core.py]: Can we update this so that it works with FlexAttention / packed sequences? Or if that's too big of a lift, I think we need to add data loading utils. Currently the way our training is configured, we don't load sequences in batches. 

I suppose we could might be able to get this running with batch_size=1, using the current data loaders. Although we will still need something custom to handle the fact that there's only 1 hidden state instead of 3.

File: src/speculators/models/fast_mtp/core.py (hunk 2 above)
Comment (fynnsu) [inline src/speculators/models/fast_mtp/core.py]: So MTP also uses train time testing?

File: src/speculators/models/fast_mtp/core.py [hunk 11]
```diff
@@ -0,0 +1,514 @@
...
+        for step in range(self.num_speculative_steps):
+            current_embeddings = self.embed_tokens(current_input_ids)
+
+            mtp_output = self.mtp_layers[0](
+                hidden_states=current_hidden,
+                token_embeddings=current_embeddings,
+                attention_mask=attention_mask,
+                position_ids=position_ids,
+                **kwargs,
+            )
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/core.py:140]: I'm concerned that the attention mask isn't correct here.

We need to have some kind of kv cache so that subsequent steps can attend to kv states from the original sequence. And we need the attention mask to account for that.

File: src/speculators/models/fast_mtp/core.py [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+
+    def load_state_dict(self, state_dict: dict, *args, **kwargs):  # type: ignore[override]
+        """Load checkpoint with automatic format detection and key remapping."""
+        sample_keys = list(state_dict.keys())[:20]
+        detected_format = self._detect_checkpoint_format(sample_keys)
+
+        if detected_format != "speculators":
+            state_dict = self._remap_checkpoint_keys(
+                state_dict, from_format=detected_format
+            )
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/core.py]: I don't think we should be handling multiple state dict structures here, just the native speculators one, which ideally would align to whatever the main existing format is.

Conversion logic like this should be implemented as a preprocessing step, before calling load_state_dict.

File: src/speculators/models/fast_mtp/mimo_config.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+__all__ = ["MiMoConfig"]
+
+
+class MiMoConfig(Qwen2Config):
+    """Transformer config for MiMo/FastMTP models.
+
+    MiMo is a Qwen2-based architecture that adds multi-token prediction (MTP)
+    fields. This shim preserves those fields when loading/saving configs.
+    """
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/mimo_config.py]: Why do we need this? Can we just use the `Qwen2Config`?

File: src/speculators/models/fast_mtp/core.py [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+    apply MLP with residual, and final normalization.
+
+    Attribute names must match checkpoint key structure exactly. The component
+    registry ensures correct module types for different architectures.
+
+    :param config: FastMTP configuration
+    :param layer_idx: Layer index (0-indexed)
+    """
+
+    def __init__(self, config: FastMTPConfig, layer_idx: int = 0):
```
Comment (fynnsu) [inline src/speculators/models/fast_mtp/core.py]: Could you comment on what's different between this and the `Eagle3FirstLayerMixin` logic?
Or what's different between this and a typical `Qwen2DecoderLayer` (or whatever type this is based on)?

File: tests/integration/models/test_fast_mtp_model.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    model = SpeculatorModel.from_pretrained(
+        model_id, verifier_attachment_mode="detached"
+    )
+    assert isinstance(model, FastMTPSpeculator)
+    config = model.config
+    assert isinstance(config, FastMTPConfig)
+
+    model.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size)
+    model.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
+    model.eval()
```
Comment (fynnsu) [inline tests/integration/models/test_fast_mtp_model.py]: Why are `embed_tokens` and `lm_head` setup here? What is the intended flow for users to set these up?

**Review by fynnsu — CHANGES_REQUESTED** [review]
Hi @rahul-tuli, thanks for the pr. I added some questions below. Can you also clarify what is the main difference between this and eagle3? I'm trying to understand if we can reduce code repetition across the two methods, since it seems like the structure is pretty similar.
